diff -up pam_krb5-2.3.11-1/src/password.c pam_krb5-2.3.11-1/src/password.c
--- pam_krb5-2.3.11-1/src/password.c	2011-10-31 17:17:20.764858601 -0400
+++ pam_krb5-2.3.11-1/src/password.c	2011-10-31 17:17:26.293897230 -0400
@@ -163,10 +163,6 @@ pam_sm_chauthtok(pam_handle_t *pamh, int
 		retval = PAM_AUTH_ERR;
 		prelim_attempted = 0;
 
-		/* Ideally we're only going to let libkrb5 ask questions once,
-		 * and after that we intend to lie to it. */
-		use_third_pass = options->use_third_pass;
-
 		/* Set up options for getting password-changing creds. */
 		i = v5_alloc_get_init_creds_opt(ctx, &tmp_gicopts);
 		if (i == 0) {
@@ -245,17 +241,6 @@ pam_sm_chauthtok(pam_handle_t *pamh, int
 			}
 		}
 		if ((password != NULL) && (i == PAM_SUCCESS)) {
-			if (options->debug) {
-				if (use_third_pass) {
-					debug("trying previously-entered "
-					      "password for '%s', allowing "
-					      "libkrb5 to prompt for more",
-					      user);
-				} else {
-					debug("trying previously-entered "
-					      "password for '%s'", user);
-				}
-			}
 			/* We have a password, so try to obtain initial
 			 * credentials using the password. */
 			i = v5_get_creds(ctx, pamh,
@@ -263,13 +248,10 @@ pam_sm_chauthtok(pam_handle_t *pamh, int
 					 options,
 					 PASSWORD_CHANGE_PRINCIPAL,
 					 password, tmp_gicopts,
-					 use_third_pass ?
-					 _pam_krb5_normal_prompter :
-					 _pam_krb5_previous_prompter,
+					 _pam_krb5_always_fail_prompter,
 					 NULL,
 					 &tmp_result);
 			prelim_attempted = 1;
-			use_third_pass = 0;
 			if (options->debug) {
 				debug("Got %d (%s) acquiring credentials for "
 				      "%s: %s.",
@@ -300,59 +282,15 @@ pam_sm_chauthtok(pam_handle_t *pamh, int
 		 * credentials using the password. */
 		if ((retval != PAM_SUCCESS) &&
 		    ((password != NULL) && (i == PAM_SUCCESS))) {
-			if (options->debug) {
-				if (use_third_pass) {
-					debug("trying newly-entered "
-					      "password for '%s', allowing "
-					      "libkrb5 to prompt for more",
-					      user);
-				} else {
-					debug("trying newly-entered "
-					      "password for '%s'", user);
-				}
-			}
 			i = v5_get_creds(ctx, pamh,
 					 &stash->v5creds, user, userinfo,
 					 options,
 					 PASSWORD_CHANGE_PRINCIPAL,
 					 password, tmp_gicopts,
-					 use_third_pass ?
-					 _pam_krb5_normal_prompter :
 					 _pam_krb5_always_fail_prompter,
 					 NULL,
 					 &tmp_result);
 			prelim_attempted = 1;
-			use_third_pass = 0;
-			if (options->debug) {
-				debug("Got %d (%s) acquiring credentials for "
-				      "%s.",
-				      tmp_result, v5_error_message(tmp_result),
-				      PASSWORD_CHANGE_PRINCIPAL);
-			}
-			retval = i;
-		}
-		/* We haven't tried anything yet, so if it's allowed, try to
-		 * obtain initial credentials, letting libkrb5 ask the
-		 * questions. */
-		if ((retval != PAM_SUCCESS) &&
-		    (prelim_attempted == 0) &&
-		    use_third_pass) {
-			if (options->debug) {
-				debug("not using an entered password for '%s', "
-				      "allowing libkrb5 to prompt", user);
-			}
-			i = v5_get_creds(ctx, pamh,
-					 &stash->v5creds, user, userinfo,
-					 options,
-					 PASSWORD_CHANGE_PRINCIPAL,
-					 NULL, tmp_gicopts,
-					 options->permit_password_callback ?
-					 _pam_krb5_always_prompter :
-					 _pam_krb5_normal_prompter,
-					 NULL,
-					 &tmp_result);
-			prelim_attempted = 1;
-			use_third_pass = 0;
 			if (options->debug) {
 				debug("Got %d (%s) acquiring credentials for "
 				      "%s.",
